### PR TITLE
Adding the definition of `nx.effective_size` for isolated nodes, and solve the division error in #6916

### LINKED ISSUE #6916: Strange behavior in `nx.effective_size` for the single node graph
Hello! It seems I've created quite a few issue reports, and I apologize for any inconvenience. Please feel free to review them at your own pace, or you can assign them to me if I can assist in creating PR to fix them.

According to the formula
image
The `effective_size` should have a well-defined value of `0` for a graph with a single node.
However, if we use Borgatti's method to calculate this graph, the value will lose its definition, and NetworkX will return some strange behavior:

For instance:
```Python
import networkx as nx

G = nx.Graph()
G.add_node(0)
print(nx.effective_size(G))

G = nx.Graph([(0, 0)])
print(nx.effective_size(G))
```
#### Result:
```
{0: nan}
ZeroDivisionError
```
I am not sure whether such behaviors are expected. So it would be highly appreciated if you could further confirm and investigate it.
If they are unexpected, I will create a PR to fix it by returning a single `0` if the graph contains only one node.

Best regards,
Joye

### Environment
NetworkX 3.1
Python 3.10

### PR DESCRIPTION
**Pull request by joyemang33**
Hi all! Sorry for being late to investigate my bug reports~ Based on our discussion, I try to fix #6916 and define `effective size` for isolated nodes by giving the value `0` as the original formula.

Please feel free to tell me if my implementation needs to improve.

Best regards,
Joye

### CODE REVIEW

**Comment by joyemang33** [comment]
Looks like we need to modify the following test cases:
image

File: networkx/algorithms/components/strongly_connected.py [hunk 1]
```diff
@@ -238,15 +238,16 @@ def strongly_connected_components_recursive(G):
     """
 
     def visit(v, cnt):
-        root[v] = cnt
-        visited[v] = cnt
-        cnt += 1
+        root[v] = cnt[0]
+        visited[v] = cnt[0]
+        cnt[0] += 1
```
Comment (dschult) [inline networkx/algorithms/components/strongly_connected.py:243]: I'm pretty sure we should remove `strongly_connected_components_recursive`. But I guess that is independent of this PR. Still, don't spend a lot of time making this recursive function work. Let's talk about that in #6898 

The reason this is showing up in this PR even though you made it for #6898 is that you are using the same branch ("main") for both PRs.  You should try to keep your main branch clean -- and up to date with the official upstream networkx repository.  Then to start each PR, make a new branch that holds your changes. add and commit your changes and then push to your githb repo and finally make the PR to the networkx repo.

File: networkx/algorithms/structuralholes.py [hunk 2]
```diff
@@ -144,17 +144,15 @@ def redundancy(G, u, v, weight=None):
     # Use Borgatti's simplified formula for unweighted and undirected graphs
     if not G.is_directed() and weight is None:
         for v in nodes:
-            # Effective size is not defined for isolated nodes
-            if len(G[v]) == 0:
-                effective_size[v] = float("nan")
+            if len(G[v]) - list(G[v]).count(v) == 0:
+                effective_size[v] = 0
```
Comment (dschult) [inline networkx/algorithms/structuralholes.py:148]: How about: (for efficiency)
```python
if len(G[v].keys() - {}) == 0:
```
or maybe (I haven't check the timing)
```python
if len(G[v]) - (v in G[v]) == 0:
```

File: networkx/algorithms/components/strongly_connected.py (hunk 1 above)
Comment (joyemang33) [reply networkx/algorithms/components/strongly_connected.py:243]: Thank you very much for your valuable review! To be honest, I'm not very familiar with these pull request features, thank you for your suggestion on that :). I will start a new branch tomorrow!

PS: +1 for removing `strongly_connected_components_recursive`.

Best regards,
Joye

File: networkx/algorithms/structuralholes.py (hunk 2 above)
Comment (joyemang33) [reply networkx/algorithms/structuralholes.py:148]: Thanks for that! I am not sure about the MultiGraph that contains multiple edges `<v, v>`, how does the behavior of these codes? I will try to check them both for functionalities and timing.

Best regards,
Joye

Comment (dschult) [reply networkx/algorithms/structuralholes.py:148]: MultiGraph has the structure node-dict of nbr-dict of edgekey-dict of edgeattribute-dict.  So `G[u]` is a dict keyed by nbr. And `G[n][nbr]` is a dict keyed by edgekey, so `len(G[n][nbr])` is the number of edges between `n` and `nbr`. For this use, the idiom `v in G[v]` returns `True` when `v` is a neighbor of `v`. And that works for either Graph or MultiGraph.  Generally we have tried to arrange it so most code written for Graph can work as expected for MultiGraph. That doesn't always work, and certainly doesn't work the other way around. :)

Comment (joyemang33) [reply networkx/algorithms/structuralholes.py:148]: It makes sense, I will try to give a code that also works well on the MultiGraph

**Comment by rossbar** [comment]
@joyemang33 are you planning on continuing with this one? If so, I think we should close this and reopen a new one that originates from a different branch (i.e. not your main branch) which should help with unintended changes (i.e. the scc stuff) and will allow reviewers to push directly to your branch, speeding up the review process.

Alternatively, if you don't have the bandwidth it's perfectly fine to close this and start fresh!

**Comment by joyemang33** [comment]
Sure! I will close it and reopen it one try to fix it with the best performance.
(Sorry for being not active recently since my laptop has been broken and just got fixed 😢 )

Best regards,
Joye

**Pull request closed without merging**